cpp/src/parquet/chunker_internal_test.cc

```diff
@@ -0,0 +1,1397 @@
+  // generate
+  size_t bytes_per_record;
+  // Data page version to use
+  ParquetDataPageVersion data_page_version = ParquetDataPageVersion::V1;
+};
+
+// Define PrintTo for MyStruct
+void PrintTo(const CaseConfig& param, std::ostream* os) {
+  *os << "{ " << param.dtype->ToString();
+  if (param.is_nullable) {
+    *os << " nullable";
+  }
+  *os << " }";
+}
+
+class TestCDC : public ::testing::Test {
+ public:
+  uint64_t GetMask(const ContentDefinedChunker& cdc) const { return cdc.GetMask(); }
+};
+
+TEST_F(TestCDC, RollingHashMaskCalculation) {
+  auto le = LevelInfo();
+  auto min_size = 256 * 1024;
+  auto max_size = 1024 * 1024;
+
+  auto cdc0 = ContentDefinedChunker(le, min_size, max_size, 0);
+  ASSERT_EQ(GetMask(cdc0), 0xFFFE000000000000);
+
+  auto cdc1 = ContentDefinedChunker(le, min_size, max_size, 1);
+  ASSERT_EQ(GetMask(cdc1), 0xFFFC000000000000);
+
+  auto cdc2 = ContentDefinedChunker(le, min_size, max_size, 2);
+  ASSERT_EQ(GetMask(cdc2), 0xFFF8000000000000);
+
+  auto cdc3 = ContentDefinedChunker(le, min_size, max_size, 3);
+  ASSERT_EQ(GetMask(cdc3), 0xFFF0000000000000);
+
+  auto cdc4 = ContentDefinedChunker(le, min_size, max_size, -1);
+  ASSERT_EQ(GetMask(cdc4), 0xFFFF000000000000);
+
+  // this is the smallest possible mask always matching, by using 8 hashtables
+  // we are going to have a match every 8 bytes; this is an unrealistic case
+  // but checking for the correctness of the mask calculation
+  auto cdc5 = ContentDefinedChunker(le, 0, 16, 0);
+  ASSERT_EQ(GetMask(cdc5), 0x0000000000000000);
+
+  auto cdc6 = ContentDefinedChunker(le, 0, 32, 1);
+  ASSERT_EQ(GetMask(cdc6), 0x0000000000000000);
+
+  auto cdc7 = ContentDefinedChunker(le, 0, 16, -1);
+  ASSERT_EQ(GetMask(cdc7), 0x8000000000000000);
+
+  // another unrealistic case, checking for the validation
+  auto cdc8 = ContentDefinedChunker(le, 128, 384, -60);
+  ASSERT_EQ(GetMask(cdc8), 0xFFFFFFFFFFFFFFFF);
+}
+
+TEST_F(TestCDC, WriteSingleColumnParquetFile) {
+  // Define the schema with a single column "number"
+  auto schema = std::dynamic_pointer_cast<schema::GroupNode>(schema::GroupNode::Make(
+      "root", Repetition::REQUIRED,
+      {schema::PrimitiveNode::Make("number", Repetition::REQUIRED, Type::INT32)}));
+
+  auto sink = CreateOutputStream();
+  auto builder = WriterProperties::Builder();
+  auto props = builder.enable_content_defined_chunking()->build();
+
+  auto writer = ParquetFileWriter::Open(sink, schema, props);
+  auto row_group_writer = writer->AppendRowGroup();
+
+  // Create a column writer for the "number" column
+  auto column_writer = row_group_writer->NextColumn();
+  auto& int_column_writer = dynamic_cast<Int32Writer&>(*column_writer);
+
+  std::vector<int32_t> numbers = {1, 2, 3, 4, 5};
+  std::vector<uint8_t> valid_bits = {1, 0, 1, 0, 1};
+  EXPECT_THROW(
+      int_column_writer.WriteBatch(numbers.size(), nullptr, nullptr, numbers.data()),
+      ParquetException);
```

Comment:
Why does it throw? Please check the exception message as well.

Reply:
See https://github.com/apache/arrow/pull/45360#discussion_r1978502781

Reply:
Does Gtest have an assertion for this? Seems like we're not asserting on the exception messages too often in the codebase.

Reply:
Yes we do. You can find `EXPECT_THROW_THAT` from the Parquet subdirectory.